source/auth/auth.rst

```diff
@@ -120,6 +120,56 @@ Drivers MUST consider a server ``Unknown`` if authentication fails. Effectively,
 	#. A driver MUST perform authentication with all supplied credentials for all server types with the exception of RSArbiter.
 	#. A single invalid credential is the same as all credentials being invalid.
 
+Mechanism Negotiation via Handshake
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:since: 4.0
+
+If a user provides a username but does not provide an authentication mechanism,
+drivers MUST issue an ``isMaster`` command requesting a user's supported SASL
+mechanisms::
+
+    {isMaster: 1, saslSupportedMechs: "<dbname>.<username>"}
+
+In this example ``<dbname>`` is the authentication database name (default
+'admin') and ``<username>`` is the username provided in the auth credential.
+The username MUST NOT be modified from the form provided by the user (i.e.  do
+not normalize with SASLprep), as the server uses the raw form to look for
+conflicts with legacy credentials.
+
+If the ``isMaster`` command fails, drivers must consider authentication to have
+failed.  If the ``isMaster`` command succeeds and the response includes a
+``saslSupportedMechs`` field, then drivers MUST use the contents of that field
+to select a default mechanism as described later.  If the command succeeds and
+the response does not include a ``saslSupportedMechs`` field, then drivers MUST
+use the legacy default mechanism rules for servers older than 4.0.
+
+Single-credential drivers
+`````````````````````````
+
+Drivers that allow only a single credential per client MUST perform mechanism
+negotiation on the initial connection handshake when the authentication
+mechanism is not specified.  This lets authentication proceed without a
+separate negotiation round-trip exchange with the server.
+
+Multi-credential drivers
+````````````````````````
+
+The use of multiple credentials within a driver is discouraged, but some
+legacy drivers still allow this.
+
+Such drivers MUST NOT attempt mechanism negotation as part of the connection
```

Comment:
This seems overly restrictive.  Drivers that support multiple credentials should have deprecated that support by now as part of the sessions work, and have added APIs that don't allow multiple credentials.  Furthermore, the majority of multiple credential API usage will be with a single credential anyway (list of one).

Given that, consider allowing drivers to treat the first credential in the credentials list the same as it would the single credential case, and then require an additional ismaster round-trip for each additional credential in the list.

Reply:
Done.